utils.go

```diff
@@ -0,0 +1,64 @@
+package main
+
+import (
+	"net"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// Returns the preferedIPProtocol, the dialProtocol, and sets the probeIPProtocolGauge.
+func configureProtocol(preferredIPProtocol string, targetAddress string, module Module, probeIPProtocolGauge prometheus.Gauge) (string, string, error) {
+	var fallbackProtocol string
+	var dialProtocol string
+
+	if preferredIPProtocol == "ip6" || preferredIPProtocol == "" {
+		preferredIPProtocol = "ip6"
+		fallbackProtocol = "ip4"
+	} else {
+		preferredIPProtocol = "ip4"
+		fallbackProtocol = "ip6"
+	}
+
+	var ip *net.IPAddr
+	var err error
+
+	if module.DNS.TransportProtocol == "udp" || module.DNS.TransportProtocol == "tcp" || module.DNS.TransportProtocol == "" {
+		ip, err = net.ResolveIPAddr(preferredIPProtocol, targetAddress)
+		if err != nil {
+			ip, err = net.ResolveIPAddr(fallbackProtocol, targetAddress)
+			if err != nil {
+				return "", "", err
+			}
+		}
+	}
+
+	if ip == nil {
+		if module.DNS.TransportProtocol[len(module.DNS.TransportProtocol)-1] == '6' {
+			probeIPProtocolGauge.Set(6)
+		} else {
+			probeIPProtocolGauge.Set(4)
+		}
+		dialProtocol = module.DNS.TransportProtocol
+		return preferredIPProtocol, dialProtocol, nil
+	}
+
+	if module.DNS.PreferredIPProtocol == "" {
+		if ip.IP.To4() == nil {
+			probeIPProtocolGauge.Set(6)
+			dialProtocol = "tcp6"
+		} else {
+			probeIPProtocolGauge.Set(4)
+			dialProtocol = "tcp4"
+		}
+		return preferredIPProtocol, dialProtocol, nil
+	}
+
+	if ip.IP.To4() == nil {
+		probeIPProtocolGauge.Set(6)
+		dialProtocol = module.DNS.TransportProtocol + "6"
```

Comment:
This is DNS specific logic, it should remain in the dns function

dialProtocol doesn't belong in here.